src/mem/epoch/markable_atomic.rs

```diff
@@ -5,33 +5,38 @@ use std::sync::atomic::{self, Ordering};
 
 use super::{Owned, Shared, Guard};
 
-/// Like `std::sync::atomic::AtomicPtr`.
+/// Provides atomic access to a `(*const T, usize)` pair.
 ///
-/// Provides atomic access to a (nullable) pointer of type `T`, interfacing with
-/// the `Owned` and `Shared` types.
+/// The pair consists of:
+///
+/// - a (nullable) pointer of type `T`, interfacing with the `Owned` and `Shared` types
+///
+/// - a `usize` marker value that can fit into the unused bits of the pointer:
+/// `mark < std::mem::align_of::<T>()`
 #[derive(Debug)]
-pub struct MarkedAtomic<T> {
+pub struct MarkableAtomic<T> {
     ptr: atomic::AtomicUsize,
     _marker: PhantomData<*const T>,
 }
 
-unsafe impl<T: Sync> Send for MarkedAtomic<T> {}
-unsafe impl<T: Sync> Sync for MarkedAtomic<T> {}
+unsafe impl<T: Sync> Send for MarkableAtomic<T> {}
+unsafe impl<T: Sync> Sync for MarkableAtomic<T> {}
 
+/// Verifies that the mark can fit into the unused bits of a pointer to `T`.
 fn guard_mark<T>(mark: usize) {
     assert!(mark < mem::align_of::<T>(),
-            "overlarge mark in MarkedAtomic: {} >= {}", mark, mem::align_of::<T>());
+            "mark too large to fit into unused bits of pointer in MarkableAtomic: {} >= {}", mark, mem::align_of::<T>());
 }
 
+/// Retrieves the original pointer and the mark from a packed value.
 fn unpack_mark<T>(val: usize) -> (*mut T, usize) {
-    let ptr = (val & !(mem::align_of::<T>() - 1)) as *mut T;
-    let mark = val &  (mem::align_of::<T>() - 1);
-    (ptr, mark)
+    let mark = val & (mem::align_of::<T>() - 1);
+    ((val - mark) as *mut T, mark)
```

Comment:
`align_of` should definitely be evaluated at compile time (if for no other reason that it, as a generic function, gets monomorphised), and I think using only bit operations to extract the parts is clearer anyway.